Mergin/utils.py

```diff
@@ -602,59 +607,112 @@ def package_layer(layer, project_dir):
-            QMessageBox.warning(None, "Error Packaging Layer", warn)
+            raise PackagingError(f"Couldn't properly save layer {layer.name()}: {err}")
 
     elif layer.type() == QgsMapLayerType.RasterLayer:
 
-        if dp.dataSourceUri().startswith("GPKG:"):
-            uri = dp.dataSourceUri()[5:]
-            dp_uri = uri[:uri.rfind(":")]
-            gpkg_table_name = uri[uri.rfind(":"):]
-        else:
-            dp_uri = dp.dataSourceUri() if os.path.isfile(dp.dataSourceUri()) else None
-            if dp_uri is None:
-                warn = f"Couldn't properly save layer: {layer.name()}\nIs it a file based layer?"
-                QMessageBox.warning(None, "Error Packaging Layer", warn)
-                return False
-
-        layer_filename = os.path.join(project_dir, os.path.basename(dp_uri))
+        save_raster_layer(layer, project_dir)
 
-        raster_writer = QgsRasterFileWriter(layer_filename)
+    else:
+        # meshes and anything else
+        raise PackagingError("Layer type not supported")
 
-        if dp.dataSourceUri().startswith("GPKG:"):
-            gdal.GetDriverByName("GPKG").Create(layer_filename, 1, 1, 1)
-            raster_writer.setOutputFormat("gpkg")
-            raster_writer.setCreateOptions([f"RASTER_TABLE={gpkg_table_name}", "APPEND_SUBDATASET=YES"])
 
-        pipe = QgsRasterPipe()
-        if not pipe.set(dp.clone()):
-            warn = f"Couldn't set raster pipe projector for layer: {layer_filename}\nSkipping..."
-            QMessageBox.warning(None, "Error Packaging Layer", warn)
-            return False
-
-        projector = QgsRasterProjector()
-        projector.setCrs(dp.crs(), dp.crs())
-        if not pipe.insert(2, projector):
-            warn = f"Couldn't set raster pipe provider for layer: {layer_filename}\nSkipping..."
-            QMessageBox.warning(None, "Error Packaging Layer", warn)
-            return False
-
-        res = raster_writer.writeRaster(pipe, dp.xSize(), dp.ySize(), dp.extent(), dp.crs())
-        if not res == QgsRasterFileWriter.NoError:
-            warn = f"Couldn't save raster: {layer_filename}\nWrite error: {res}"
-            QMessageBox.warning(None, "Error Packaging Layer", warn)
-            return False
+def save_raster_layer(raster_layer, project_dir):
+    """
+    Save raster layer to the project directory.
+    If the source raster is a GeoTiff, create a copy of the original file in the project directory.
+    If it is a GeoPackage raster, save it also as a GeoPackage table.
+    Otherwise, save the raster as GeoTiff using Qgs with some optimisations.
+    """
 
-        provider_opts = QgsDataProvider.ProviderOptions()
-        provider_opts.layerName = layer.name()
-        datasource = layer_filename
-        layer.setDataSource(datasource, layer.name(), "gdal", provider_opts)
+    driver_name = get_raster_driver_name(raster_layer)
 
+    if driver_name == "GTiff":
+        copy_tif_raster(raster_layer, project_dir)
```

Comment:
Just wondering - does packaging work if the source raster is a COG hosted on some remote server? (we don't need to support that use case, but it should not fail with a python error - instead we should just disable packaging of remote COGs if that's not supported)

Reply:
Added a check for COGs - if they are remote the packaging option for them is disabled and they can be either skipped of kept as they are. In the future we can add also packaging of remote rasters (and possibly db-stored too).